File: android/src/main/kotlin/com/retozu/flutter_amazon_chime/EventAnalyticsObserver.kt
/*
 * Copyright (c) 2025 Duong Le. MIT License.
 */

package com.retozu.flutter_amazon_chime

import com.amazonaws.services.chime.sdk.meetings.analytics.EventAnalyticsObserver
import com.amazonaws.services.chime.sdk.meetings.analytics.EventAttributes
import com.amazonaws.services.chime.sdk.meetings.analytics.EventName

class ChimeEventAnalyticsObserver(private val chimeFlutterApi: ChimeFlutterApi) : EventAnalyticsObserver {
    override fun onEventReceived(name: EventName, attributes: EventAttributes) {
        val stringAttributes: Map<String?, Any?> = attributes.entries.associate { (key, value) -> key.name to value }
        chimeFlutterApi.onMeetingEventReceived(name.name, stringAttributes) {}
    }
}

File: ios/Classes/EventAnalyticsObserver.swift
import AmazonChimeSDK

class ChimeEventAnalyticsObserver: EventAnalyticsObserver {
    let chimeFlutterApi: ChimeFlutterApi

    init(chimeFlutterApi: ChimeFlutterApi) {
        self.chimeFlutterApi = chimeFlutterApi
    }

    func eventDidReceive(name: EventName, attributes: [AnyHashable: Any]) {
        var stringAttributes = [String: Any?]()
        for (key, value) in attributes {
            stringAttributes[String(describing: key)] = toPigeonValue(value)
        }
        chimeFlutterApi.onMeetingEventReceived(name: String(describing: name), attributes: stringAttributes) { _ in }
    }

    private func toPigeonValue(_ value: Any) -> Any? {
        switch value {
        case let s as String: return s
        case let n as NSNumber: return n
        case let b as Bool: return b
        default: return String(describing: value)
        }
    }
}
